compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -860,6 +860,11 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
             && let hir::Node::Expr(lhs) = self.tcx.hir_node(*lhs_hir_id)
             && let hir::Node::Expr(rhs) = self.tcx.hir_node(*rhs_hir_id)
             && let Some(rhs_ty) = typeck_results.expr_ty_opt(rhs)
+            && let trait_pred = predicate.unwrap_or(trait_pred)
+            && hir::lang_items::OPERATORS.iter().any(|&op| {
+                // Ensure we only run this code on operators
+                self.tcx.require_lang_item(op, None) == trait_pred.skip_binder().trait_ref.def_id
+            })
```

Comment:
Diagnostic code should not _require_ language items. Please use a fallible API.

```suggestion
            // Ensure we only run this code on operators
            && hir::lang_items::OPERATORS.iter().filter_map(|&op| self.tcx.lang_items().get(op)).any(|op| {
                op == trait_pred.skip_binder().trait_ref.def_id
            })
```

@rustbot author

Reply:
This needs to check that it's a binary operator as well. This still would ICE if I had a nested unary operator as a predicate.

Reply:
> Diagnostic code should not require language items. Please use a fallible API.

Fixed, thanks for pointing this out.

> This still would ICE if I had a nested unary operator as a predicate.

I don't think this is necessarily the case--we already check that the rhs is `Some`, which would fail if it were a unary op: https://github.com/sjwang05/rust/blob/issue-119352/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs#L856

@rustbot review

Reply:
> I don't think this is necessarily the case--we already check that the rhs is Some

No, you can have a binary operator predicate with a unary operator predicate as a `where` clause bound.

Reply:
Ahh, I see. I fixed the check to filter out unary ops.

@rustbot review